File: 500.600/558.LogicOrOfTwoBinaryGridsRepresentedAsQuadTrees.cpp
#include <iostream>
using namespace std;

class Node
{
public:
    bool val{};
    bool isLeaf{};
    Node *topLeft{};
    Node *topRight{};
    Node *bottomLeft{};
    Node *bottomRight{};

    Node() {}

    Node(bool _val, bool _isLeaf) : val(_val), isLeaf(_isLeaf)
    {
    }

    Node(bool _val, bool _isLeaf, Node *_topLeft, Node *_topRight, Node *_bottomLeft, Node *_bottomRight)
        : val(_val), isLeaf(_isLeaf), topLeft(_topLeft), topRight(_topRight), bottomLeft(_bottomLeft), bottomRight(_bottomRight)
    {
    }
};

class LogicOrOfTwoBinaryGridsRepresentedAsQuadTrees
{
public:
    Node *intersect(Node *quadTree1, Node *quadTree2)
    {
        if (quadTree1->isLeaf)
        {
            if (quadTree1->val)
            {
                return new Node(true, true);
            }
            return quadTree2;
        }
        if (quadTree2->isLeaf)
        {
            return intersect(quadTree2, quadTree1);
        }
        Node *o1 = intersect(quadTree1->topLeft, quadTree2->topLeft);
        Node *o2 = intersect(quadTree1->topRight, quadTree2->topRight);
        Node *o3 = intersect(quadTree1->bottomLeft, quadTree2->bottomLeft);
        Node *o4 = intersect(quadTree1->bottomRight, quadTree2->bottomRight);
        if (o1->isLeaf && o2->isLeaf && o3->isLeaf && o4->isLeaf && o1->val == o2->val && o1->val == o3->val && o1->val == o4->val)
        {
            return new Node(o1->val, true);
        }
        return new Node(false, false, o1, o2, o3, o4);
    }
};

void del(Node *head)
{
    if (head->topLeft)
    {
        del(head->topLeft);
    }
    if (head->topRight)
    {
        del(head->topRight);
    }
    if (head->bottomLeft)
    {
        del(head->bottomLeft);
    }
    if (head->bottomRight)
    {
        del(head->bottomRight);
    }
    delete head;
}

void print(Node *head)
{
    cout << head->val << " , ";
    if (head->topLeft)
    {
        print(head->topLeft);
    }
    if (head->topRight)
    {
        print(head->topRight);
    }
    if (head->bottomLeft)
    {
        print(head->bottomLeft);
    }
    if (head->bottomRight)
    {
        print(head->bottomRight);
    }
    delete head;
}

int main()
{
    LogicOrOfTwoBinaryGridsRepresentedAsQuadTrees cls;

    Node *head1 = new Node(1, 0);
    head1->topLeft = new Node(1, 1);
    head1->topRight = new Node(1, 1);
    head1->bottomLeft = new Node(0, 1);
    head1->bottomRight = new Node(0, 1);
    Node *head2 = new Node(1, 0);
    head2->topLeft = new Node(1, 1);
    head2->topRight = new Node(1, 0);
    head2->bottomLeft = new Node(1, 1);
    head2->bottomRight = new Node(0, 1);
    head2->topRight->topLeft = new Node(0, 1);
    head2->topRight->topRight = new Node(0, 1);
    head2->topRight->bottomLeft = new Node(1, 1);
    head2->topRight->bottomRight = new Node(1, 1);
    Node *ret = cls.intersect(head1, head2);
    print(ret);
    cout << endl;
    del(ret);
    return 0;
}

File: go/500.600/558.LogicOrOfTwoBinaryGridsRepresentedAsQuadTrees.go
package main

import "fmt"

type Node struct {
	Val         bool
	IsLeaf      bool
	TopLeft     *Node
	TopRight    *Node
	BottomLeft  *Node
	BottomRight *Node
}

func (t *Node) print() {
	fmt.Printf("[%v,%v]", t.Val, t.IsLeaf)
}

func intersect(quadTree1 *Node, quadTree2 *Node) *Node {
	if quadTree1.IsLeaf {
		if quadTree1.Val {
			return &Node{Val: true, IsLeaf: true}
		}
		return quadTree2
	}
	if quadTree2.IsLeaf {
		return intersect(quadTree2, quadTree1)
	}
	o1 := intersect(quadTree1.TopLeft, quadTree2.TopLeft)
	o2 := intersect(quadTree1.TopRight, quadTree2.TopRight)
	o3 := intersect(quadTree1.BottomLeft, quadTree2.BottomLeft)
	o4 := intersect(quadTree1.BottomRight, quadTree2.BottomRight)
	if o1.IsLeaf && o2.IsLeaf && o3.IsLeaf && o4.IsLeaf && o1.Val == o2.Val && o1.Val == o3.Val && o1.Val == o4.Val {
		return &Node{Val: o1.Val, IsLeaf: true}
	}
	return &Node{Val: false, IsLeaf: false, TopLeft: o1, TopRight: o2, BottomLeft: o3, BottomRight: o4}
}

func print(h *Node) {
	h.print()
	if h.TopLeft != nil {
		print(h.TopLeft)
	}
	if h.TopRight != nil {
		print(h.TopRight)
	}
	if h.BottomLeft != nil {
		print(h.BottomLeft)
	}
	if h.BottomRight != nil {
		print(h.BottomRight)
	}
}

func main() {
	head1 := &Node{Val: true, IsLeaf: false}
	head1.TopLeft = &Node{Val: true, IsLeaf: true}
	head1.TopRight = &Node{Val: true, IsLeaf: true}
	head1.BottomLeft = &Node{Val: false, IsLeaf: true}
	head1.BottomRight = &Node{Val: false, IsLeaf: true}
	head2 := &Node{Val: true, IsLeaf: false}
	head2.TopLeft = &Node{Val: true, IsLeaf: true}
	head2.TopRight = &Node{Val: true, IsLeaf: false}
	head2.BottomLeft = &Node{Val: true, IsLeaf: true}
	head2.BottomRight = &Node{Val: false, IsLeaf: true}
	head2.TopRight.TopLeft = &Node{Val: false, IsLeaf: true}
	head2.TopRight.TopRight = &Node{Val: false, IsLeaf: true}
	head2.TopRight.BottomLeft = &Node{Val: true, IsLeaf: true}
	head2.TopRight.BottomRight = &Node{Val: true, IsLeaf: true}
	ret := intersect(head1, head2)
	print(ret)
}
